### ISSUE TITLE
Write GameStudio logs to a temporary folder

### ISSUE THREAD

**Issue by manio143** [stride3d/stride#950]
**Is your feature request related to a problem? Please describe.**
A few times it happened that a person had an issue running Stride, even if they were shown a log, maybe they closed the window and couldn't provide more details.

**Describe the solution you'd like**
We should save all the logs into a temporary file so that it can be inspected later.

**Additional context**
I took a look at current log listeners and we can easily add an extension to the `TextWriterLogListener`:
```csharp
public class TempFileLogListener : TextWriterLogListener
{
    public TempFileLogListener(string prefix)
        :base(File.OpenWrite(Path.Combine(Path.GetTempPath(), "Stride", $"{prefix}_{System.DateTime.Now:yyyyMMdd_Hmmss}.log")))
    {
    }
}
```
And then attach it when we create logger instances (discussion can be held which of those should be kept in the file):
https://github.com/stride3d/stride/blob/1c8daf98112556c4256e3c900dd6a23fd6ce6920/sources/editor/Stride.GameStudio/Program.cs#L142
https://github.com/stride3d/stride/blob/1c8daf98112556c4256e3c900dd6a23fd6ce6920/sources/shared/Stride.NuGetResolver/NuGetAssemblyResolver.cs#L76
☝ These two is the most important ones - a lot of things uses the GlobalLogger and a lot of problems can occur during NuGet resolution.
https://github.com/stride3d/stride/blob/3eacea405fe7f2c890e859e4fe14a18efd086982/sources/editor/Stride.GameStudio/DebuggingViewModel.cs#L78
https://github.com/stride3d/stride/blob/273dfddd462fd3746569f833e1493700c070b14d/sources/editor/Stride.Assets.Presentation/AssetEditors/GameEditor/Services/EditorGameController.cs#L99
https://github.com/stride3d/stride/blob/273dfddd462fd3746569f833e1493700c070b14d/sources/editor/Stride.Core.Assets.Editor/ViewModel/AssetCollectionViewModel.cs#L481
https://github.com/stride3d/stride/blob/273dfddd462fd3746569f833e1493700c070b14d/sources/editor/Stride.Core.Assets.Editor/ViewModel/AssetSourcesViewModel.cs#L117
https://github.com/stride3d/stride/blob/922ba8bdaa3ea4d2ca9be288a5b5df6a0e48e3bf/sources/editor/Stride.Core.Assets.Editor/ViewModel/SessionViewModel.cs#L252
https://github.com/stride3d/stride/blob/273dfddd462fd3746569f833e1493700c070b14d/sources/editor/Stride.Core.Assets.Editor/ViewModel/AssetSourceTrackerViewModel.cs#L95

There is a lot of small loggerResult's that aren't listened to all over the codebase.

**Comment by thargy** [comment]
As temporary files are not easily discovered, due to the randomness of the name, there has to be some form of use.  How would the user/developed, know where to look for crash data.

The most obvious use, to me, would be to offer to include the temporary file (post flushing) to any error crash report.

**Comment by manio143** [comment]
If you at the piece of code I wrote there it would create files in `%TEMP%/Stride/{prefix}_{datetime}.log` where prefix can signify whether it was the GameStudio global logger or something else. So it would be discoverable. For crash reporting - I'm not sure what is the actual infrastructure for receiving reports, but the way it works right now is it stores the last X warnings/errors in a cyclic buffer and includes that with the fatal exception. Temp logging could be more verbose than that.

**Comment by thargy** [comment]
Yes, sorry I skimmed your code and just assumed you'd used `Path.GetTempFileName()` for some inexplicable reason.  Nonetheless, temp files are still not super discoverable, as you'd have to know the framework was logging there.  That said, it would be possible to ask for someone to look for the file and send it.

I believe `Stride.Editor.CrashReport` runs in process on unhandled exceptions, and that creates a set of data that can be serialized to JSON.  It would be relatively trivial to look in the temp folder for the latest log file and ask the user whether they wish to include it.  Although, `CrashReportData` doesn't is from an external NuGet for which the source doesn't appear to be publically available.

However, as you rightly point out, it dumps the log ring buffer into the crash report anyway, which is actually more likely to be up to date at that point.  The log file is only really useful after the process has terminated, but the crash report system doesn't appear to be using a watchdog model but is instead called as the process is terminating with an unhandled exception.  That's the least reliable form of crash reporting though.

I had erroneously assumed a watchdog was used as there's already a launcher, so it's quite easy to monitor the launched process.  The most reliable approach though (in my experience) is to check for incorrectly terminated applications on the next run.  This is often not thought of as it lacks the immediacy and people worry that it will mean missed logs (people will see a crash and never run it again).  In practice, that is rarely the case as users almost always run it at least once more after a crash, and almost always immediately after it crashes.

This is approach is really easy to implement, especially if we implement file logs, and is actually the most reliable crash detection technique (it even 'detects' power cuts!).  One of the simplest tricks is to terminate the process's log on a valid exit, with an 'end marker'.  On startup, you check the last log for that termination, and if absent you prompt the user to send a crash report, asking them what they were doing (you can even not send the report if they select 'there was a power cut' :rofl:).  You can also then delete the previous log file after confirming it was terminated correctly, preventing filing a disk, and start a new log.

For this reason, it wouldn't be bad to use JSON as a log format.  Not least as that can more readily be added quickly to the existing crash report logic (which I believe uses JSON).  Moreover, if the logs are stored in an array, and the array is only terminated on process exit, the file would effectively be invalid JSON until terminated, and an unterminated JSON file would be easy to very quick to detect on startup (technically you can just seek the final non-whitespace character and check it's a `]`, which is super fast, even on huge log files (assuming the array contents are objects).

So I would recommend we store the logs as JSON, and terminate the array on a valid process exit.  This will allow for a potential future change to the crash reporting mechanism to add detection of non-terminated logs on startup and to generate a crash report where the logs are read from the file in that scenario.

**Comment by frankhaugen** [comment]
Wouldn't an implementation of ILogger<> from Microsoft.Extensions.Logging be cool? At least I'd like the same kind of "extensebility", so that I could supply my own logging provider without changing implementation anywhere
